// @thirdweb-dev/react/dist/magicLink-71cb3272.cjs.prod.js
'use strict';

var wallets = require('@thirdweb-dev/wallets');
var styled = require('@emotion/styled');
var reactIcons = require('@radix-ui/react-icons');
var React = require('react');
var reactCore = require('@thirdweb-dev/react-core');
var basic = require('./basic-5133a200.cjs.prod.js');
var jsxRuntime = require('react/jsx-runtime');
require('@emotion/react');
require('@radix-ui/react-dialog');
require('@radix-ui/colors');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var styled__default = /*#__PURE__*/_interopDefault(styled);

const EmailConnect = props => {
  const [isConnecting, setIsConnecting] = React.useState(false);
  const createInstance = reactCore.useCreateWalletInstance();
  const twContext = reactCore.useThirdwebWallet();
  const [email, setEmail] = React.useState("");
  const [isValidEmail, setIsValidEmail] = React.useState(false);
  const handleConnect = async () => {
    const magicWallet = createInstance(props.magicLinkWallet);
    setIsConnecting(true);
    props.close();
    const connectOptions = {
      chainId: twContext.activeChain?.chainId,
      email
    };
    try {
      twContext.setConnectionStatus("connecting");
      await magicWallet.connect(connectOptions);
      setIsConnecting(false);
      twContext.handleWalletConnect(magicWallet, connectOptions);
      props.onConnect();
    } catch (e) {
      console.error(e);
      setIsConnecting(false);
      twContext.setConnectionStatus("disconnected");
      props.open();
    }
  };
  const error = email && !isValidEmail;
  return /*#__PURE__*/jsxRuntime.jsxs(jsxRuntime.Fragment, {
    children: [/*#__PURE__*/jsxRuntime.jsx(basic.BackButton, {
      onClick: props.onBack
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
      y: "md"
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Img, {
      src: props.magicLinkWallet.meta.iconURL,
      width: basic.iconSize.xl,
      height: basic.iconSize.xl
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
      y: "md"
    }), /*#__PURE__*/jsxRuntime.jsx(basic.ModalTitle, {
      children: " Login with email "
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
      y: "lg"
    }), /*#__PURE__*/jsxRuntime.jsxs("form", {
      onSubmit: e => {
        e.preventDefault();
        handleConnect();
      },
      children: [/*#__PURE__*/jsxRuntime.jsx(basic.Input, {
        type: "email",
        id: "email",
        name: "email",
        variant: "outline",
        placeholder: "johndoe@gmail.com",
        required: true,
        value: email,
        onChange: e => {
          setEmail(e.target.value);
          // check if a valid phone number using browser api
          setIsValidEmail(e.target.validity.valid);
        },
        "data-error": error
      }), /*#__PURE__*/jsxRuntime.jsxs("div", {
        style: {
          visibility: error ? "visible" : "hidden"
        },
        children: [/*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
          y: "sm"
        }), /*#__PURE__*/jsxRuntime.jsx(basic.ErrorMessage, {
          children: "Invalid Email "
        })]
      }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
        y: "lg"
      }), /*#__PURE__*/jsxRuntime.jsx(basic.FormFooter, {
        children: /*#__PURE__*/jsxRuntime.jsxs(basic.Button, {
          variant: "inverted",
          style: {
            gap: basic.spacing.sm
          },
          children: [isConnecting ? "Signing in" : "Sign in", isConnecting && /*#__PURE__*/jsxRuntime.jsx(basic.Spinner, {
            color: "inverted",
            size: "sm"
          })]
        })
      })]
    })]
  });
};

const SMSConnect = props => {
  const [isConnecting, setIsConnecting] = React.useState(false);
  const createInstance = reactCore.useCreateWalletInstance();
  const twContext = reactCore.useThirdwebWallet();
  const [phoneNumber, setPhoneNumber] = React.useState("");
  const [isValidPhoneNumber, setIsValidPhoneNumber] = React.useState(false);
  const handleSmsConnect = async () => {
    const magicWallet = createInstance(props.magicLinkWallet);
    setIsConnecting(true);
    props.close();
    const connectOptions = {
      chainId: twContext.activeChain?.chainId,
      phoneNumber
    };
    try {
      twContext.setConnectionStatus("connecting");
      await magicWallet.connect(connectOptions);
      setIsConnecting(false);
      twContext.handleWalletConnect(magicWallet, connectOptions);
      props.onConnect();
    } catch (e) {
      console.error(e);
      setIsConnecting(false);
      twContext.setConnectionStatus("disconnected");
      props.open();
    }
  };
  const error = phoneNumber && !isValidPhoneNumber;
  return /*#__PURE__*/jsxRuntime.jsxs(jsxRuntime.Fragment, {
    children: [/*#__PURE__*/jsxRuntime.jsx(basic.BackButton, {
      onClick: props.onBack
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
      y: "md"
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Img, {
      src: props.magicLinkWallet.meta.iconURL,
      width: basic.iconSize.xl,
      height: basic.iconSize.xl
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
      y: "md"
    }), /*#__PURE__*/jsxRuntime.jsx(basic.ModalTitle, {
      children: " Login with phone number "
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
      y: "sm"
    }), /*#__PURE__*/jsxRuntime.jsx(basic.ModalDescription, {
      children: "Enter your phone number including country code"
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
      y: "lg"
    }), /*#__PURE__*/jsxRuntime.jsxs("form", {
      onSubmit: e => {
        e.preventDefault();
        handleSmsConnect();
      },
      children: [/*#__PURE__*/jsxRuntime.jsx(basic.Input, {
        type: "tel",
        id: "phone",
        name: "phone",
        variant: "outline",
        placeholder: "+1234567890",
        required: true,
        value: phoneNumber,
        pattern: "\\+[0-9]{1}[0-9]+",
        onChange: e => {
          setPhoneNumber(e.target.value);
          // check if a valid phone number using browser api
          setIsValidPhoneNumber(e.target.validity.valid);
        },
        "data-error": error
      }), /*#__PURE__*/jsxRuntime.jsxs("div", {
        style: {
          visibility: error ? "visible" : "hidden"
        },
        children: [/*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
          y: "sm"
        }), /*#__PURE__*/jsxRuntime.jsx(basic.ErrorMessage, {
          children: !phoneNumber.startsWith("+") ? "Phone number must start with +" : "Invalid phone number"
        })]
      }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
        y: "lg"
      }), /*#__PURE__*/jsxRuntime.jsx(basic.FormFooter, {
        children: /*#__PURE__*/jsxRuntime.jsxs(basic.Button, {
          variant: "inverted",
          style: {
            gap: basic.spacing.sm
          },
          children: [isConnecting ? "Connecting" : "Connect", isConnecting && /*#__PURE__*/jsxRuntime.jsx(basic.Spinner, {
            color: "inverted",
            size: "sm"
          })]
        })
      })]
    })]
  });
};

const MagicConnectUI = props => {
  const isSmsEnabled = props.magicWallet.config.smsLogin !== false;
  const isEmailEnabled = props.magicWallet.config.emailLogin !== false;
  let firstScreen = "menu";
  if (isEmailEnabled && !isSmsEnabled) {
    firstScreen = "email";
  } else if (isSmsEnabled && !isEmailEnabled) {
    firstScreen = "sms";
  }

  // const createInstance = useCreateWalletInstance();
  // const twContext = useThirdwebWallet();
  // const [isConnecting, setIsConnecting] = useState(false);
  const [showUI, setShowUI] = React.useState(firstScreen);

  // const handleAuthConnect = async (oauthProvider: MagicOAuthProvider) => {
  //   const magicWallet = createInstance(magicLinkObj) as MagicLink;
  //   setIsConnecting(true);
  //   await magicWallet.connect({
  //     oauthProvider: oauthProvider,
  //   });
  //   setIsConnecting(false);
  //   twContext.handleWalletConnect(magicWallet);
  //   props.onConnect();
  // };

  if (showUI === "sms") {
    return /*#__PURE__*/jsxRuntime.jsx(SMSConnect, {
      close: props.close,
      open: props.open,
      magicLinkWallet: props.magicWallet,
      onBack: () => {
        if (firstScreen === "sms") {
          props.goBack();
        } else {
          setShowUI(firstScreen);
        }
      },
      onConnect: props.close
    });
  }
  if (showUI === "email") {
    return /*#__PURE__*/jsxRuntime.jsx(EmailConnect, {
      close: props.close,
      open: props.open,
      magicLinkWallet: props.magicWallet,
      onBack: () => {
        if (firstScreen === "email") {
          props.goBack();
        } else {
          setShowUI(firstScreen);
        }
      },
      onConnect: props.close
    });
  }

  // if (isConnecting) {
  //   return (
  //     <Flex
  //       justifyContent="center"
  //       alignItems="center"
  //       style={{
  //         height: "350px",
  //       }}
  //     >
  //       <Spinner color="primary" size="lg" />
  //     </Flex>
  //   );
  // }

  // const authButton = (oauthProvider: MagicOAuthProvider) => (
  //   <IconButton
  //     variant="secondary"
  //     onClick={() => handleAuthConnect(oauthProvider)}
  //     aria-label={oauthProvider}
  //   >
  //     <Img
  //       src={authProviderImages[oauthProvider]}
  //       width={iconSize.lg}
  //       height={iconSize.lg}
  //     />
  //   </IconButton>
  // );

  return /*#__PURE__*/jsxRuntime.jsxs(jsxRuntime.Fragment, {
    children: [/*#__PURE__*/jsxRuntime.jsx(basic.BackButton, {
      onClick: props.goBack
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
      y: "md"
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Img, {
      src: props.magicWallet.meta.iconURL,
      width: basic.iconSize.xl,
      height: basic.iconSize.xl
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
      y: "md"
    }), /*#__PURE__*/jsxRuntime.jsx(basic.ModalTitle, {
      children: " Magic Link "
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
      y: "sm"
    }), /*#__PURE__*/jsxRuntime.jsxs(basic.ModalDescription, {
      children: [isSmsEnabled && isEmailEnabled && "Login with your phone number or email", isSmsEnabled && !isEmailEnabled && "Login with your phone number", !isSmsEnabled && isEmailEnabled && "Login with your email"]
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
      y: "xl"
    }), /*#__PURE__*/jsxRuntime.jsxs(basic.Flex, {
      flexDirection: "column",
      gap: "sm",
      children: [isSmsEnabled && /*#__PURE__*/jsxRuntime.jsxs(LoginButton, {
        variant: "secondary",
        onClick: () => {
          setShowUI("sms");
        },
        children: ["Phone number", /*#__PURE__*/jsxRuntime.jsx(reactIcons.ChatBubbleIcon, {
          width: basic.iconSize.md,
          height: basic.iconSize.md,
          style: {
            marginLeft: "auto"
          }
        })]
      }), isEmailEnabled && /*#__PURE__*/jsxRuntime.jsxs(LoginButton, {
        variant: "secondary",
        onClick: () => {
          setShowUI("email");
        },
        children: ["Email", /*#__PURE__*/jsxRuntime.jsx(reactIcons.EnvelopeClosedIcon, {
          width: basic.iconSize.md,
          height: basic.iconSize.md,
          style: {
            marginLeft: "auto"
          }
        })]
      }), !isEmailEnabled && !isSmsEnabled && /*#__PURE__*/jsxRuntime.jsx(basic.ErrorMessage, {
        children: " No login methods enabled "
      })]
    }), /*#__PURE__*/jsxRuntime.jsx(basic.Spacer, {
      y: "xl"
    })]
  });
};
const LoginButton = styled__default["default"](basic.Button)`
  width: 100%;
  gap: ${basic.spacing.md};
  justify-content: left;
  &:hover {
    background-color: ${p => p.theme.bg.elevatedHover};
  }
`;

// const authProviderImages: Record<MagicOAuthProvider, string> = {
//   google: "ipfs://QmNMm6313vpMxbyTcXyZMSEVMkpTvkmJXaqCyFrM5TDQpV/google.svg",
//   twitter: "ipfs://QmbUePooAWJbY1ZbzamAb36WJkuNCDzKxuFh4ZUbrepLud/twitter.svg",
//   apple: "",
//   bitbucket: "",
//   discord: "",
//   facebook: "",
//   github:
//     "ipfs://QmVN2Jaz4XGpEdqo9Ki16TJmxxhiiXDaG4kP18uXpGrJQA/github-gray.svg",
//   gitlab: "",
//   linkedin: "",
//   microsoft:
//     "ipfs://QmX4Nk5VW4tJVFstnEgXvZBdx5uVDz3Si662nhVNpWUFai/microsoft.svg",
//   twitch: "",
// };

const magicLink = config => {
  const configuredWallet = {
    id: wallets.MagicLink.id,
    meta: wallets.MagicLink.meta,
    create: options => new wallets.MagicLink({
      ...options,
      ...config
    }),
    config,
    connectUI(props) {
      return /*#__PURE__*/jsxRuntime.jsx(MagicConnectUI, {
        ...props,
        magicWallet: configuredWallet
      });
    },
    isInstalled() {
      return false;
    }
  };
  return configuredWallet;
};

exports.magicLink = magicLink;
