// @thirdweb-dev/wallets/dist/thirdweb-dev-wallets.esm.js
export { a as AbstractClientWallet, A as AsyncLocalStorage, D as DEFAULT_DAPP_META, c as createAsyncLocalStorage } from './base-0142d377.esm.js';
export { w as walletIds } from './walletIds-93948c58.esm.js';
export { C as CredentialsStorage } from './CredentialsStorage-88c8666b.esm.js';
export { W as WagmiConnector } from './WagmiConnector-3965c5ad.esm.js';
export { A as AddChainError, C as ChainNotConfiguredError, P as ProviderRpcError, S as SwitchChainError, U as UserRejectedRequestError } from './errors-161bdcae.esm.js';
export { n as normalizeChainId } from './normalizeChainId-dd5a6036.esm.js';
export { C as Connector, W as WagmiAdapter } from './connector-887e68b5.esm.js';
export { a as assertWindowEthereum } from './assertWindowEthereum-bb2b4544.esm.js';
export { PaperWallet } from '../evm/wallets/paper-wallet/dist/thirdweb-dev-wallets-evm-wallets-paper-wallet.esm.js';
export { CoinbaseWallet } from '../evm/wallets/coinbase-wallet/dist/thirdweb-dev-wallets-evm-wallets-coinbase-wallet.esm.js';
export { LocalWallet, isValidPrivateKey } from '../evm/wallets/local-wallet/dist/thirdweb-dev-wallets-evm-wallets-local-wallet.esm.js';
export { InjectedWallet } from '../evm/wallets/injected/dist/thirdweb-dev-wallets-evm-wallets-injected.esm.js';
export { MetaMaskWallet } from '../evm/wallets/metamask/dist/thirdweb-dev-wallets-evm-wallets-metamask.esm.js';
export { WalletConnect } from '../evm/wallets/wallet-connect/dist/thirdweb-dev-wallets-evm-wallets-wallet-connect.esm.js';
export { WalletConnectV1 } from '../evm/wallets/wallet-connect-v1/dist/thirdweb-dev-wallets-evm-wallets-wallet-connect-v1.esm.js';
export { SafeSupportedChainsSet } from '../evm/connectors/safe/dist/thirdweb-dev-wallets-evm-connectors-safe.esm.js';
export { SafeWallet } from '../evm/wallets/safe/dist/thirdweb-dev-wallets-evm-wallets-safe.esm.js';
export { MagicLink } from '../evm/wallets/magic/dist/thirdweb-dev-wallets-evm-wallets-magic.esm.js';
export { SmartWallet, getAllSigners, getAllSmartWallets, getSmartWalletAddress, isSmartWalletDeployed } from '../evm/wallets/smart-wallet/dist/thirdweb-dev-wallets-evm-wallets-smart-wallet.esm.js';
export { EthersWallet } from '../evm/wallets/ethers/dist/thirdweb-dev-wallets-evm-wallets-ethers.esm.js';
export { PrivateKeyWallet } from '../evm/wallets/private-key/dist/thirdweb-dev-wallets-evm-wallets-private-key.esm.js';
import './classPrivateMethodGet-71fe23d8.esm.js';
import './classPrivateFieldSet-e25e1cec.esm.js';
import './defineProperty-e24c82ea.esm.js';
import '../evm/wallets/abstract/dist/thirdweb-dev-wallets-evm-wallets-abstract.esm.js';
import 'ethers';
import 'eventemitter3';
import '@thirdweb-dev/chains';
import 'buffer';
import '@safe-global/safe-ethers-adapters';
import '@safe-global/safe-core-sdk';
import '@safe-global/safe-ethers-lib';
import '@thirdweb-dev/sdk';
import './keys-5e5ba251.esm.js';
