// @thirdweb-dev/wallets/dist/thirdweb-dev-wallets.browser.esm.js
export { a as AbstractClientWallet, A as AsyncLocalStorage, D as DEFAULT_DAPP_META, c as createAsyncLocalStorage } from './base-e1301995.browser.esm.js';
export { w as walletIds } from './walletIds-1ea97411.browser.esm.js';
export { C as CredentialsStorage } from './CredentialsStorage-aad81d2c.browser.esm.js';
export { W as WagmiConnector } from './WagmiConnector-60b19604.browser.esm.js';
export { A as AddChainError, C as ChainNotConfiguredError, P as ProviderRpcError, S as SwitchChainError, U as UserRejectedRequestError } from './errors-105ad187.browser.esm.js';
export { n as normalizeChainId } from './normalizeChainId-e4cc0175.browser.esm.js';
export { C as Connector, W as WagmiAdapter } from './connector-c1e1817a.browser.esm.js';
export { a as assertWindowEthereum } from './assertWindowEthereum-66ccef46.browser.esm.js';
export { PaperWallet } from '../evm/wallets/paper-wallet/dist/thirdweb-dev-wallets-evm-wallets-paper-wallet.browser.esm.js';
export { CoinbaseWallet } from '../evm/wallets/coinbase-wallet/dist/thirdweb-dev-wallets-evm-wallets-coinbase-wallet.browser.esm.js';
export { LocalWallet, isValidPrivateKey } from '../evm/wallets/local-wallet/dist/thirdweb-dev-wallets-evm-wallets-local-wallet.browser.esm.js';
export { InjectedWallet } from '../evm/wallets/injected/dist/thirdweb-dev-wallets-evm-wallets-injected.browser.esm.js';
export { MetaMaskWallet } from '../evm/wallets/metamask/dist/thirdweb-dev-wallets-evm-wallets-metamask.browser.esm.js';
export { WalletConnect } from '../evm/wallets/wallet-connect/dist/thirdweb-dev-wallets-evm-wallets-wallet-connect.browser.esm.js';
export { WalletConnectV1 } from '../evm/wallets/wallet-connect-v1/dist/thirdweb-dev-wallets-evm-wallets-wallet-connect-v1.browser.esm.js';
export { SafeSupportedChainsSet } from '../evm/connectors/safe/dist/thirdweb-dev-wallets-evm-connectors-safe.browser.esm.js';
export { SafeWallet } from '../evm/wallets/safe/dist/thirdweb-dev-wallets-evm-wallets-safe.browser.esm.js';
export { MagicLink } from '../evm/wallets/magic/dist/thirdweb-dev-wallets-evm-wallets-magic.browser.esm.js';
export { SmartWallet, getAllSigners, getAllSmartWallets, getSmartWalletAddress, isSmartWalletDeployed } from '../evm/wallets/smart-wallet/dist/thirdweb-dev-wallets-evm-wallets-smart-wallet.browser.esm.js';
export { EthersWallet } from '../evm/wallets/ethers/dist/thirdweb-dev-wallets-evm-wallets-ethers.browser.esm.js';
export { PrivateKeyWallet } from '../evm/wallets/private-key/dist/thirdweb-dev-wallets-evm-wallets-private-key.browser.esm.js';
import './classPrivateMethodGet-ea199cc3.browser.esm.js';
import './classPrivateFieldSet-a5db7c83.browser.esm.js';
import './defineProperty-c8ecdc07.browser.esm.js';
import '../evm/wallets/abstract/dist/thirdweb-dev-wallets-evm-wallets-abstract.browser.esm.js';
import 'ethers';
import 'eventemitter3';
import '@thirdweb-dev/chains';
import 'buffer';
import '@safe-global/safe-ethers-adapters';
import '@safe-global/safe-core-sdk';
import '@safe-global/safe-ethers-lib';
import '@thirdweb-dev/sdk';
import './keys-bd983fca.browser.esm.js';
